[SourceKit] Improvements to stamp handling in SwiftASTManager

ahoppen opened the pull request:
Some refactoring of `SwiftASTManager`, motivated by the fact that we were computing an input file’s realpath 3 times from `fileContentsForFilesInCompilerInvocation`.

ahoppen committed:
[SourceKit] Remove some dupliate computation of realpaths

ahoppen committed:
[SourceKit] Remove ability to specify fixed snapshots to use when building an AST

The feature was never used and just complicated the implementation.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

bnbarham commented on tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp:
```diff
@@ -889,15 +877,11 @@ ASTBuildOperation::snapshotAndStampsForFilesInCompilerInvocation(
        Opts.Invok.getFrontendOptions().InputsAndOutputs.getAllInputs()) {
     const std::string &Filename = input.getFileName();
     bool IsPrimary = input.isPrimary();
-    if (auto FoundSnapshot = findSnapshot(Snapshots, Filename)) {
```
Can we remove `findSnapshot` now?

ahoppen replied:
Yes, we do. Removed it.

bnbarham commented on tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp:
```diff
@@ -888,35 +876,27 @@ ASTBuildOperation::fileContentsAndStampsForFilesInCompilerInvocation() {
-  std::vector<BufferStamp> InputStamps;
-  InputStamps.reserve(
-      Opts.Invok.getFrontendOptions().InputsAndOutputs.inputCount());
   // IMPORTANT: The computation of stamps must match the one in
   // snapshotAndStampsForFilesInCompilerInvocation.
```
This function doesn't exist any more

ahoppen replied:
Removed the comment. It was also out of date.

ahoppen committed:
[SourceKit] Improve memory cost function of ASTBuildOperation

ahoppen committed:
[SourceKit] Store FileContents in ASTBuildOperation instead of snapshots

ahoppen committed:
[SourceKit] Don't store Stamps in ASTBuildOperation

FileContents already contains the stamps.

ahoppen committed:
[SourceKit] Make FileContent non ref-counted

We never need to have two copies of the same `FileContent` object, so we don’t need a copy constructor and can thus pass it on the stack again, instead of storing it on the heap.

ahoppen commented:
@swift-ci Please smoke test
